### ISSUE TITLE
Exposing 3 pods in a stateful set under the same domain name, while allowing specific pod targeting by clients

### ISSUE THREAD

**Issue by salehio** [Azure/AKS#1928]
TL;DR: I can't find a good way to expose 3 pods in a stateful set under the same domain name (for SSL) while still allowing clients to be able to target pods individually.

**What happened**:
I'm running a stateful set (neo4j causal cluster) in my AKS cluster and I want to expose it to the internet.

Neo4j is a horizontally scalable graph database. It exposes itself over the bolt protocol normally on port 7687. Neo4j also supports SSL with certificates.

Writers to the database also need to be able to target specific pods/instances, since only one of them can be written to at once.
First the client requests the addresses (**domain or IP** + port) of all the existing pods from any one of the pods, and then it routes read/write requests accordingly.

Since I have 3 pods for my neo4j instances in my stateful set, I decided to:
1- Create 3 IPs
2- Map 7687 to the pod based on pod-name via load balancer
<pre>
<code>
apiVersion: v1
kind: Service
metadata:
  labels:
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: neo4j
    statefulset.kubernetes.io/pod-name: neo4j-core-$INDEX
  name: neo4j-external-$INDEX
spec:
  loadBalancerIP: $IP <b># This is different depending on $INDEX</b>
  externalTrafficPolicy: Cluster
  ports:
  - name: bolt
    port: 7687
    protocol: TCP
    targetPort: 7687
  selector:
   <b>statefulset.kubernetes.io/pod-name: neo4j-core-$INDEX</b>
  sessionAffinity: None
  type: LoadBalancer

</code>
</pre>
3- Map my domain to the first IP. (A record for mycooldomain.com -> IpAddresses[0])

So now I have 3 services, and the 3 IPs are mapping from 7687 externally to the internal pods on the correct port.
My client can now talk to mycooldomain.com, which resolves to IpAddresses[0], and it makes a call to get IpAddresses[1] and IpAddresses[2], and everything is nice and dandy.

Once I turn on SSL, things started failing. The nodes were advertising themselves based on IP (remember, the list of IPs is fetched on the first request). This doesn't match the SSL certificate.

Since I only have 1 cert, that means I need to expose all nodes under the same domain
Because the clients need to target specific pods, the only way I could think of to achieve that while using the same domain was to change the ports.
Have IpAddress[0]:768**7** map to statefulset.kubernetes.io/pod-name: neo4j-core-**0**
Have IpAddress[0]:768**8** map to statefulset.kubernetes.io/pod-name: neo4j-core-**1**
Have IpAddress[0]:768**9** map to statefulset.kubernetes.io/pod-name: neo4j-core-**2**

and to also configure pods to expose themselves via domain name. That means on first request, the client will fetch the following addresses:
mycooldomain.com:768**7**
mycooldomain.com:768**8**
mycooldomain.com:768**9**

which would resolve correctly based on the selector.

I could not find a way to have selectors over individual pods in AKS, so I needed to create 3 separate Service YAML manifests and deploy them. They have different names, of course.
I modify the load balancer template above to use the same IP regardless of value of $INDEX, and to also have unique external ports mapping to 7687 on the pod, then I redeploy.
After deploying, I notice things are not working.

Checking the error on the service we see this:
<pre>
<code>
Events:
  Type     Reason                  Age                  From                Message
  ----     ------                  ----                 ----                -------
  Normal   EnsuringLoadBalancer    56s (x6 over 3m40s)  service-controller  Ensuring load balancer
  Warning  SyncLoadBalancerFailed  55s (x6 over 3m38s)  service-controller  Error syncing load balancer: failed to ensure load balancer: Retriable: false, RetryAfter: 0s, HTTPStatusCode: 400, RawError: Retriable: false, RetryAfter: 0s, HTTPStatusCode: 400, RawError: {
  "error": {
    "code": "PublicIPReferencedByMultipleIPConfigs",
    "message": "Public ip address /subscriptions/X/resourceGroups/Y/providers/Microsoft.Network/publicIPAddresses/Z is referenced by multiple ipconfigs in resource /subscriptions/X/resourceGroups/Y/providers/Microsoft.Network/loadBalancers/kubernetes.",
    "details": []
  }
}
</code>
</pre>

I can't really find a way to achieve this today. How can I expose 3 pods over SSL with hostname verification while still allowing the ability to target individual pods with requests?
It seems like I need a way to have multiple selectors per service based on port.
